### PR TITLE
Re-create file system when the GCS token changes

### PR DESCRIPTION
**Pull request by beinan**
Just like what we did for hdfs and kerberos,  we should also re-create/refresh the file-system in PrestoFileSystemCache when the GCS access token changes.

### CODE REVIEW

File: src/main/java/org/apache/hadoop/fs/PrestoFileSystemCache.java [hunk 1]
```diff
@@ -219,6 +238,12 @@ private static boolean isHdfs(URI uri)
         return "hdfs".equals(scheme) || "viewfs".equals(scheme);
     }
 
+    private static boolean isGcs(URI uri)
```
Comment (zhenxiao) [inline src/main/java/org/apache/hadoop/fs/PrestoFileSystemCache.java:241]: isGCS?

File: src/main/java/org/apache/hadoop/fs/PrestoFileSystemCache.java [hunk 2]
```diff
@@ -219,6 +238,12 @@ private static boolean isHdfs(URI uri)
...
+    {
+        String scheme = uri.getScheme();
```
Comment (zhenxiao) [inline src/main/java/org/apache/hadoop/fs/PrestoFileSystemCache.java:243]: inline the implementation:
"gs".equals(uri.getScheme())

File: src/main/java/org/apache/hadoop/fs/PrestoFileSystemCache.java [hunk 3]
```diff
@@ -105,6 +106,24 @@ private synchronized FileSystem getInternal(URI uri, Configuration conf, long un
         return fileSystemHolder.getFileSystem();
     }
 
+    private boolean shouldFileSystemHolderRefresh(URI uri, Configuration conf, Set<?> privateCredentials, FileSystemHolder fileSystemHolder)
+    {
+        if (isHdfs(uri)) {
+            return !fileSystemHolder.getPrivateCredentials().equals(privateCredentials);
+        }
+        else if (isGcs(uri)) {
```
Comment (zhenxiao) [inline src/main/java/org/apache/hadoop/fs/PrestoFileSystemCache.java:114]: else is not needed, right?

File: src/main/java/org/apache/hadoop/fs/PrestoFileSystemCache.java [hunk 4]
```diff
@@ -105,6 +106,24 @@ private synchronized FileSystem getInternal(URI uri, Configuration conf, long un
...
+            String existingGcsToken = fileSystemHolder.getFileSystem().getConf().get(PRESTO_GCS_OAUTH_ACCESS_TOKEN_KEY);
+            String newGcsToken = conf.get(PRESTO_GCS_OAUTH_ACCESS_TOKEN_KEY);
+            if (existingGcsToken == null) {
+                return newGcsToken != null;
+            }
+            else {
```
Comment (zhenxiao) [inline src/main/java/org/apache/hadoop/fs/PrestoFileSystemCache.java:120]: remove else

**Comment by beinan** [comment]
@zhenxiao many thanks for the review!  all code review issues fixed.

**Comment by beinan** [comment]
Hey @zhenxiao thank you so much for the review!  do you happen to know when will be the next release of presto-hadoop-apache2 ?  We have a couple of changes on presto repo depending on this.

**Comment by zhenxiao** [comment]
@caithagoras do you know when could we do a release for presto-hadoop-apache2 ? We have a couple of changes on presto repo depending on this. CC @beinan

**Comment by caithagoras** [comment]
Since I don't have context here, cc @jainxrohit to confirm that this is compatible with with Presto at Facebook.

**Comment by jainxrohit** [comment]
> Since I don't have context here, cc @jainxrohit to make sure that this is compatible with with Presto at Facebook.
> cc @zhenxiao We can make a release when I get a confirmation from @jainxrohit .

@caithagoras Looks good to me.

**Pull request merged**